certstore_linux.go

```diff
@@ -1,14 +1,336 @@
+	var certificates = make([]*x509.Certificate, 0)
+	for node = C.CertListHead(list); C.CertListEnd(node, list) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*node)
+		identities = append(identities, &identity)
+	}
+	certificates = append(certificates, cert)
+	found := true
+	for found != false {
+		found = false
+		cert = certificates[len(certificates)-1]
+		issuer := cert.Issuer.String()
+		for i := 0; i < len(identities); i++ {
+			cert, err = identities[i].Certificate()
+			if cert == nil {
+				return nil, fmt.Errorf("Error %d, cannot fetch certificate...\n", int(C.PR_GetError()))
+			}
+			subject := cert.Subject.String()
+			root := certificates[len(certificates)-1].Subject.String()
+			if subject == issuer && subject != root {
+				certificates = append(certificates, cert)
+				found = true
+				break
+			}
+		}
+	}
+	return certificates, nil
+}
+
+func (i *nssIdentity) Delete() error {
+	C.PK11_DeleteTokenCertAndKey(i.cert, nil)
+	return nil
+}
+
+func (nssIdentity) Close() {
+}
+
+func (i *nssIdentity) Public() (crypto.PublicKey) {
+	var cert, _ = i.Certificate()
+	if cert == nil {
+		return nil
+	}
+	return cert.PublicKey
+}
+
+func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	hash := opts.HashFunc()
+	if len(digest) != hash.Size() {
+		return nil, errors.New("bad digest for hash")
+	}
+	key := C.PK11_FindKeyByAnyCert(i.cert, nil)
+	if key == nil {
+		return nil, errors.New("cannot find private key")
+	}
+	if key.keyType == C.rsaKey {
+		var pkcs1Prefix = map[crypto.Hash][]byte{
+			crypto.SHA1:   {0x30, 0x21, 0x30, 0x09, 0x06, 0x05, 0x2b, 0x0e, 0x03, 0x02, 0x1a, 0x05, 0x00, 0x04, 0x14},
+			crypto.SHA256: {0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20},
+			crypto.SHA384: {0x30, 0x41, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0x04, 0x30},
+			crypto.SHA512: {0x30, 0x51, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0x04, 0x40},
+		}
+		oid := pkcs1Prefix[hash]
+		if oid == nil {
+			return nil, ErrUnsupportedHash
+		}
+		T := make([]byte, len(oid)+len(digest))
+		copy(T[0:len(oid)], oid)
+		copy(T[len(oid):], digest)
+		digest = T
+	}
+	sd := C.SECITEM_AllocItem(nil, nil, C.uint(C.PK11_SignatureLen(key)))
+	hashed := C.SECITEM_AllocItem(nil, nil, C.uint(len(digest)))
+	if hashed == nil {
+		return nil, errors.New("SECITEM_AllocItem failed")
+	}
+	C.memcpy(unsafe.Pointer(hashed.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+	var mechanism C.ulong
+	switch key.keyType {
+	case C.rsaKey:
+		switch hash {
+		/*
+		case crypto.SHA1:
+			mechanism = C.CKM_SHA1_RSA_PKCS
+		case crypto.SHA256:
+			mechanism = C.CKM_SHA256_RSA_PKCS
+		case crypto.SHA384:
+			mechanism = C.CKM_SHA384_RSA_PKCS
+		case crypto.SHA512:
+			mechanism = C.CKM_SHA512_RSA_PKCS
+		*/
+		default:
+			mechanism = C.CKM_RSA_PKCS
```

Comment:
I'm rusty enough on PKCS11 that I don't understand what the correct behavior is here. Does `CKM_SHA1_RSA_PKCS` imply that the input isn't already digested whereas `CKM_RSA_PKCS` implies that it is already digested?

Reply:
This code is now gone